pbr-basic/main.zig

```diff
@@ -42,6 +43,216 @@ const Material = struct {
     params: Params,
 };
 
+/// The imported model contains an array of vectors that are either positions or normals. Due to the fact that
+/// we combine positions and normals into our `Vertex` structure, we need to "repack" all the indices so
+/// that they reference a contigious array of Vertex value (Containing both the position and normal)
+/// VertexIndexer takes the original "sparse" index (As indices to normals are ignored) and returns a unique index for
+/// this new packed / contingious space.
+const VertexIndexer = struct {
+    const null_index: u32 = std.math.maxInt(u32);
+    const null_normal = Vec3{
+        std.math.floatMax(f32),
+        std.math.floatMax(f32),
+        std.math.floatMax(f32),
+    };
+
+    const Record = struct {
+        normal: Vec3,
+        next: u32,
+
+        pub inline fn isNull(self: @This()) bool {
+            return std.mem.eql(f32, &self.normal, &null_normal); // self.normal0] == std.math.floatMax(f32);
+        }
+    };
+
+    const Result = struct {
+        index: u32,
+        new_vertex: bool,
+    };
+
+    buffer: []Record,
+
+    /// Map index from sparse -> packed
+    /// Sparse indices are used to fast lookup of stored normal values, but
+    /// need to be converted to indices that reference a packed buffer of vertices
+    index_map: []u32,
+
+    /// Next index outside of the 1:1 mapping range for storing
+    /// position -> normal collisions
+    next_collision_index: u32,
+
+    /// Next packed index
+    next_packed_index: u32,
+
+    pub fn init(allocator: std.mem.Allocator, collision_index_start: u32, capacity: usize) !@This() {
+        var result = VertexIndexer{
+            .buffer = try allocator.alloc(Record, capacity),
+            .index_map = try allocator.alloc(u32, capacity),
+            .next_collision_index = collision_index_start,
+            .next_packed_index = 0,
+        };
+        std.mem.set(Record, result.buffer, Record{ .normal = null_normal, .next = null_index });
+        return result;
+    }
+
+    pub fn deinit(self: *@This(), allocator: std.mem.Allocator) void {
+        allocator.free(self.buffer);
+        allocator.free(self.index_map);
+    }
+
+    pub inline fn nextRecord(self: @This(), record: Record) ?Record {
+        return if (record.next == null_index) null else self.buffer[record.next];
+    }
+
+    pub inline fn indexFor(self: *@This(), sparse_index: u32, normal: Vec3) Result {
```

Comment:
This is a pretty nifty approach if I understand it correctly!

So this function looks up a `sparse_index` (i.e. the indices we get for vertices only from model3d) and returns a "repacked" index where our unique vertex+normal is located.

`buffer[sparse_index]` stores a "chain" for every vertex. Obviously, some `sparse_index` will collide (same `sparse_index`, but a different `normal` value); and that's where the chain comes in: if `buffer[sparse_index]` already has a chain, then we walk that chain and compare our normal value with the ones stored in that chain. If it exists, we found our "repacked" vertex (this is where deduplication happens) - otherwise, we append to the chain.

Is my understanding of the logic behind this right?

---

If my understanding is correct, then I think I see a few possible improvements:

1. We could generalize this by having it take a generic `Vertex` type (so it can store normals, or normals+UVs, etc. per-vertex.)
2. It seems possible to eliminate the intermediate buffers here:
  1. Get rid of `vertices_buffer` and `indices_buffer` at the usage site, and instead move construction of those buffers into this algorithm.
  2. Replace  `buffer: []Record,` in this algorithm with two arrays: (1) `vertices: []Vertex` and (2) a `next_vertices: []u32` mapping `vertices[x] -> vertices[y]` (replacing `Record.next` 

Neither of these need be done here/by you, just writing this out as an idea for future self :)

Reply:
Yes, that's correct. 

1. Of course, should be an easy change.

As for the other changes, I hadn't actually thought of that and seems like a good idea. We'd just need to translate sparse index -> packed index to find the location of the normal, etc but we already have that mapping. If the entry doesn't exist that just means it's a new chain. I'll look into it now

Reply:
Updated to manage index & vertex buffers internally. I think it's much cleaner this way and easier to explain. I also renamed to VertexWriter. Let me know what you think.